lib/commands/recordscreen.js

```diff
@@ -1,61 +1,157 @@
+    }
+
+    const localPort = this.opts.remotePort;
+    if (this.opts.usePortForwarding) {
+      await this.startIproxy(localPort);
+    }
+
+    const args = [
+      '-f', 'mjpeg',
+      '-r', this.opts.videoFps,
+      '-i', `http://localhost:${localPort}`,
+      '-vcodec', 'mjpeg',
+      '-y', this.videoPath,
+    ];
+    this.mainProcess = new SubProcess(FFMPEG_BINARY, args);
+    this.mainProcess.on('output', (stdout, stderr) => {
+      if (stderr && !stderr.includes('frame=')) {
+        log.warn(`[${FFMPEG_BINARY}] ${stderr}`);
+      }
+    });
+    // Give ffmpeg some time for init
+    await this.mainProcess.start(5000);
+    log.info(`Starting screen capture on the device '${this.udid}' with command: '${FFMPEG_BINARY} ${args.join(' ')}'. ` +
+      `Will timeout in ${timeout}ms`);
+
+    this.timeoutHandler = setTimeout(async () => {
+      if (!await this.interrupt()) {
+        log.warn(`Cannot finish the active screen recording on the device '${this.udid}' after ${timeout}ms timeout`);
+      }
+    }, timeout);
+  }
+
+  async startIproxy (localPort) {
+    this.iproxy = new iProxy(this.udid, localPort, this.opts.remotePort);
+    try {
+      await this.iproxy.start();
+    } catch (err) {
+      log.warn(`Cannot start iproxy. Assuming it is already forwarding the remote port ${this.opts.remotePort} to ${localPort} ` +
+        `for the device ${this.udid}. Set the custom value to 'mjpegServerPort' capability if this is an undesired behavior. ` +
+        `Original error: ${err.message}`);
+      this.iproxy = null;
+    }
+  }
+
+  async stopIproxy () {
+    if (!this.iproxy) {
+      return;
+    }
+
+    const quitPromise = this.iproxy.quit();
+    this.iproxy = null;
+    try {
+      await quitPromise;
+    } catch (err) {
+      log.warn(`Cannot stop iproxy. Original error: ${err.message}`);
+    }
+  }
+
+  async terminate () {
+    let result = true;
+
+    if (this.timeoutHandler) {
+      clearTimeout(this.timeoutHandler);
+      this.timeoutHandler = null;
+    }
+
+    if (this.mainProcess && this.mainProcess.isRunning) {
       try {
-        await exec('kill', ['-0', pid]);
-      } catch (ign) {
-        return true;
+        await this.mainProcess.stop('SIGTERM');
+      } catch (e) {
+        log.warn(`Cannot terminate ${FFMPEG_BINARY}. Original error: ${e.message}`);
+        result = false;
       }
-      return false;
-    }, {waitMs: PROCESS_SHUTDOWN_TIMEOUT_SEC * 1000, intervalMs: 300});
-  } catch (e) {
-    return false;
+      this.mainProcess = null;
+    }
+
+    if (this.opts.usePortForwarding) {
+      await this.stopIproxy();
+    }
+
+    return result;
   }
-  return true;
-}
 
-async function uploadRecordedMedia (localFile, remotePath = null, uploadOptions = {}) {
-  try {
-    return await encodeBase64OrUpload(localFile, remotePath, uploadOptions);
-  } finally {
-    await fs.rimraf(localFile);
+  async interrupt () {
```

Comment:
the code for this looks mostly the same as for `terminate`, perhaps it could be shared?

Reply:
the code here is different due to the same reason as described above